keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,535 @@
+        Action core_v1.Handler
+
+        // ActionTimeoutSeconds defines when the execution hook controller should stop retrying.
+        // If execution fails, the execution hook controller will keep retrying until reaching
+        // ActionTimeoutSeconds. If execution still fails or hangs, execution hook controller
+        // stops retrying and updates executionhook status to failed.
+        // If controller loses its state, counter restarts. In this case, controller will retry
+        // for at least this long, before stopping.
+        // Once an action is started, controller has no way to stop it even if
+        // ActionTimeoutSeconds is exceeded. This simply controls if retry happens or not.
+        // retry is based on exponential backoff policy. If ActionTimeoutSeconds is not
+        // specified, it will retry until the hook object is deleted.
+        // +optional
+        ActionTimeoutSeconds *int64
+}
+```
+
+`ExecutionHook` may be used for different cases, such as
+
+* Application-consistency snapshotting
+* Upgrade
+* Rolling upgrade
+* Debugging
+* Prepare for some lifecycle event like a database migration
+* Reload a config file
+* Restart a container
+
+
+The following gives an example of how to use ExecutionHook for application-consistency snapshotting use case.
+
+### Workflow
+
+* Snapshot API should carry two execution hook information (which will reference two HookAction - preSnapshotExecutionHook and postSnapshotExecutionHook) for quiescing and unquiescing application.
+* Snapshot controller watches Snapshot objects. If there is a request to create a Snapshot, it checks if there is ExecutionHook Information defined in Snapshot.
+* If hook is defined in the snapshot, snapshot controller will create one or multiple ExecutionHooks for quiescing application if necessary, one for each command running in pods/containers.
+* Snapshot controller waits for all Action hooks to complete running before taking snapshot.
+* The ExecutionHook controller watches the ExecutionHook API object and take actions based on the object status and also update the status. 
+* Snapshot controller waits until hook is run on all pods (if more than 1 pod). No matter the Action succeeds or fails, snapshot controller should create execution hooks for unquiescing application. Snapshot controller can also decide when to delete those hooks.
+
+Here is an example of an HookAction:
+```
+apiVersion: apps.k8s.io/v1alpha1
+kind: HookAction
+metadata:
+  name: action-demo
+Action:
+  exec:
+    command: [“run_quiesce.sh”]
+  actionTimeoutSeconds: 10
+```
+
+Here is an ExecutionHook created by the snapshot controller:
+```
+apiVersion: apps.k8s.io/v1alpha1
+kind: ExecutionHook
+metadata:
+  name: hook-demo
+spec:
+  podContainerNamesList:
+    -podName: myPod1
+      -containerName: myContainer1
+      -containerName: myContainer2
+    -podName: myPod2
+      -containerName: myContainer3
+      -containerName: myContainer4
+  hookActionName: action-demo
```

Comment:
why is the hookAction a different object ? What do we loose if we combined the hookaction inside ExecutionHook ?